bip78/src/uri.rs

```diff
@@ -145,13 +146,79 @@ impl From<InternalPjParseError> for ParseUriError {
 
 #[cfg(test)]
 mod tests {
-    use crate::Uri;
+    use crate::{Uri, ParseUriError};
     use std::str::FromStr;
+    use crate::uri::{InternalBip21Error, InternalPjParseError};
+    use crate::bitcoin::util;
+    use crate::bitcoin::util::amount::ParseAmountError;
 
     #[test]
     fn test_short() {
         assert!(Uri::from_str("").is_err());
         assert!(Uri::from_str("bitcoin").is_err());
         assert!(Uri::from_str("bitcoin:").is_err());
     }
+
+
+    #[test]
+    fn test_valid() {
+        for pj in ["https://example.com", "http://example.com", "http://vjdpwgybvubne5hda6v4c5iaeeevhge6jvo3w2cl6eocbwwvwxp7b7qd.onion"].iter() {
+            let pj = format!("bitcoin:12c6DSiU4Rq3P4ZxziKxzrL5LmMBrzjrJX?amount=20.3&pj={}", pj);
+            assert!(pj.parse::<Uri>().is_ok());
+        }
+
+        assert!("bitcoin:TB1Q6D3A2W975YNY0ASUVD9A67NER4NKS58FF0Q8G4?amount=0.0001&pj=https://testnet.demo.btcpayserver.org/BTC/pj".parse::<Uri>().is_ok());
+
+        //TODO we may want endpoint to be a valid URL
+        assert!(Uri::from_str("bitcoin:TB1Q6D3A2W975YNY0ASUVD9A67NER4NKS58FF0Q8G4?amount=1&pj=http://a").is_ok());
+
+        let uri = Uri::from_str("bitcoin:TB1Q6D3A2W975YNY0ASUVD9A67NER4NKS58FF0Q8G4?amount=1&pj=http://example.com?ciao").unwrap();
+        assert_eq!(uri.endpoint, "http://example.com?ciao");
+
+    }
+
+    #[test]
+    fn test_errors() {
+        assert_eq!(
+            Uri::from_str("bitcoin:175tWpb8K1S7NmH4Zx6rewF9WQrcZv245W"),
+            Err::<Uri<'_>, ParseUriError>(ParseUriError::PjNotPresent)
+        );
+
+        assert_eq!(
+            Uri::from_str("bitcoinz:175tWpb8K1S7NmH4Zx6rewF9WQrcZv245W"),
+            Err::<Uri<'_>, ParseUriError>(
+                InternalBip21Error::BadSchema(
+                    "bitcoinz:175tWpb8K1S7NmH4Zx6rewF9WQrcZv245W".to_string()
+                )
+                    .into()
+            )
+        );
+
+        assert_eq!(
+            Uri::from_str("bitcoin:175tWpb8K1S7NmH4Zx6rewF9WQrcZv245W?amount=20.3&label=Luke-Jr"),
+            Err::<Uri<'_>, ParseUriError>(
+                InternalBip21Error::Address(util::address::Error::Base58(util::base58::Error::BadChecksum(291609738, 694262922))).into()
+            )
+        );
+
+        assert_eq!(
+            Uri::from_str("bitcoin:TB1Q6D3A2W975YNY0ASUVD9A67NER4NKS58FF0Q8G4?pj=https://testnet.demo.btcpayserver.org/BTC/pj"),
+            Err::<Uri<'_>, ParseUriError>(InternalPjParseError::MissingAmount.into())
```

Comment:
FYI we have this wrong (see e-mail from Nicolas).

Reply:
moved wrong tests in specific ignored test